codex/sales.nim

```diff
@@ -181,37 +177,39 @@ proc cleanUp(
 
   await sales.remove(agent)
 
-  # signal back to the slot queue to cycle a worker
-  if not processing.isNil and not processing.finished():
-    processing.complete()
-
-proc filled(
-    sales: Sales, request: StorageRequest, slotIndex: uint64, processing: Future[void]
-) =
+proc filled(sales: Sales, request: StorageRequest, slotIndex: uint64) =
   if onSale =? sales.context.onSale:
     onSale(request, slotIndex)
 
-  # signal back to the slot queue to cycle a worker
-  if not processing.isNil and not processing.finished():
-    processing.complete()
-
-proc processSlot(sales: Sales, item: SlotQueueItem, done: Future[void]) =
+proc processSlot(
+    sales: Sales, item: SlotQueueItem
+) {.async: (raises: [CancelledError]).} =
   debug "Processing slot from queue", requestId = item.requestId, slot = item.slotIndex
 
   let agent =
     newSalesAgent(sales.context, item.requestId, item.slotIndex, none StorageRequest)
 
+  let completed = newAsyncEvent()
```

Comment:
If this is single use (which it looks like), perhaps a simple `Future` is simpler here, `AsyncEvent` semantincs might get confusing at times. But just a suggestion :)

Reply:
I think that dance is warranted since we want to prevent it from being fired twice.

Reply:
> I think that dance is warranted since we want to prevent it from being fired twice.

The AsyncEvent ensures that we don't go through the `await completed.wait()` call more than once, even if you call `completed.fire()` more than once.